File: src/services/billing.ts
import apiClient from "@/api/apiClient";

export interface BillingPlan {
	id: number;
	name: string;
	description: string | null;
	price: number;
	features: string[];
	is_active: boolean;
	created_at: string;
	updated_at: string | null;
}

export interface UserSubscription {
	id: number;
	user_id: number;
	plan_id: number;
	status: string;
	start_date: string;
	end_date: string;
	created_at: string;
	updated_at: string | null;
	plan: BillingPlan;
}

export interface BillingHistory {
	id: number;
	user_id: number;
	plan_id: number;
	amount: number;
	status: string;
	payment_date: string;
	created_at: string;
	plan: BillingPlan;
}

export interface UsageStats {
	id: number;
	user_id: number;
	api_calls_used: number;
	api_calls_limit: number;
	storage_used: number;
	storage_limit: number;
	team_members_used: number;
	team_members_limit: number;
	created_at: string;
	updated_at: string | null;
}

export interface CurrentPlan {
	plan: BillingPlan;
	subscription: UserSubscription;
	usage_stats: UsageStats;
}

// Get all billing plans
export async function getBillingPlans() {
	return apiClient.get<BillingPlan[]>({
		url: "/billing/plans",
	});
}

// Get a specific billing plan
export async function getBillingPlan(planId: number) {
	return apiClient.get<BillingPlan>({
		url: `/billing/plans/${planId}`,
	});
}

// Get current user's plan, subscription, and usage stats
export async function getCurrentPlan() {
	return apiClient.get<CurrentPlan>({
		url: "/billing/current-plan",
	});
}

// Subscribe to a plan
export async function subscribeToPlan(planId: number) {
	return apiClient.post<UserSubscription>({
		url: `/billing/subscribe/${planId}`,
		// This endpoint will:
		// 1. Cancel any existing active subscription
		// 2. Create a new subscription for the selected plan
		// 3. Update usage limits based on the new plan
	});
}

// Get billing history
export async function getBillingHistory() {
	return apiClient.get<BillingHistory[]>({
		url: "/billing/history",
	});
}

// Get usage stats
export async function getUsageStats() {
	return apiClient.get<UsageStats>({
		url: "/billing/usage",
	});
}

File: backend/app/models/billing.py
import json
from sqlalchemy import Column, Integer, String, Float, DateTime, ForeignKey, Boolean, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from sqlalchemy.types import TypeDecorator, TEXT

from app.db.database import Base


class JSONEncodedList(TypeDecorator):
    """Represents a list as a JSON-encoded string."""

    impl = TEXT

    def process_bind_param(self, value, dialect):
        if value is not None:
            value = json.dumps(value)
        return value

    def process_result_value(self, value, dialect):
        if value is not None:
            value = json.loads(value)
        return value


class BillingPlan(Base):
    __tablename__ = "billing_plans"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    description = Column(String)
    price = Column(Float, nullable=False)
    features = Column(JSONEncodedList, nullable=False)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())


class UserSubscription(Base):
    __tablename__ = "user_subscriptions"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    plan_id = Column(Integer, ForeignKey("billing_plans.id"), nullable=False)
    status = Column(String, nullable=False, default="active")  # active, canceled, expired
    start_date = Column(DateTime(timezone=True), nullable=False)
    end_date = Column(DateTime(timezone=True), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    user = relationship("User", back_populates="subscription")
    plan = relationship("BillingPlan")


class BillingHistory(Base):
    __tablename__ = "billing_history"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    plan_id = Column(Integer, ForeignKey("billing_plans.id"), nullable=False)
    amount = Column(Float, nullable=False)
    status = Column(String, nullable=False)  # paid, failed, pending
    payment_date = Column(DateTime(timezone=True), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    user = relationship("User", back_populates="billing_history")
    plan = relationship("BillingPlan")


class UsageStats(Base):
    __tablename__ = "usage_stats"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    api_calls_used = Column(Integer, default=0)
    api_calls_limit = Column(Integer, nullable=False)
    storage_used = Column(Integer, default=0)  # in MB
    storage_limit = Column(Integer, nullable=False)  # in MB
    team_members_used = Column(Integer, default=0)
    team_members_limit = Column(Integer, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    user = relationship("User", back_populates="usage_stats") 